core/src/main/scala/kafka/server/metadata/ZkMetadataCache.scala

```diff
@@ -315,7 +324,35 @@ class ZkMetadataCache(brokerId: Int, metadataVersion: MetadataVersion, brokerFea
     }.getOrElse(Map.empty[Int, Node])
   }
 
-  def getControllerId: Option[Int] = metadataSnapshot.controllerId
+  def getControllerId: Option[Int] = {
+    val snapshot = metadataSnapshot
+    snapshot.controllerId.orElse(snapshot.kraftControllerId)
+  }
+
+  /**
+   * Return controller id. Additionally returns if the controller is zk or kraft controller.
+   * Returns true for zk controller, false otherwise.
+   * @return
+   */
+  def getZkOrKRaftControllerId: (Option[Int], Boolean) = {
+    val snapshot = metadataSnapshot
+    snapshot.kraftControllerId match {
+      case Some(controller) => (Some(controller), false)
+      case None => (snapshot.controllerId, true)
+    }
+  }
+
+  override def getControllerIdForExternalClient: Option[Int] = {
```

Comment:
Can you add a javadoc here explaining the behavior (i.e., random alive broker)?

Reply:
Added to the `MetadataCache` trait.